### PR TITLE
Allow including YAML files

### PR DESCRIPTION
**Pull request by lebauce**
When using multiple scenarios with many platforms, we either need to duplicate the entries in every scenario or provide a base config. The latter does not really provide fine-grained granularity. Being able to include YAML files allows to share only some specific sections between scenarios.

#### PR Type

- Feature Pull Request

### CODE REVIEW

File: src/molecule/config.py [hunk 1]
```diff
@@ -416,9 +418,11 @@ def _get_defaults(self) -> MutableMapping:
             },
         }
 
-    def _preflight(self, data: MutableMapping):
+    def _preflight(self, data: MutableMapping, include_path: str = ""):
```
Comment (decentral1se) [inline src/molecule/config.py:421]: Can that be a `pathlib.Path`? Not sure.

File: src/molecule/model/schema_v3.py [hunk 2]
```diff
@@ -367,9 +367,9 @@ def _validate_molecule_env_var(self, molecule_env_var, field, value):
                 self._error(field, msg)
 
 
-def pre_validate(stream, env: MutableMapping, keep_string: str):
+def pre_validate(stream, env: MutableMapping, keep_string: str, include_path=""):
```
Comment (decentral1se) [inline src/molecule/model/schema_v3.py:370]: Missing type declaration like you included elsewhere?

File: src/molecule/util.py [hunk 3]
```diff
@@ -224,29 +255,50 @@ def safe_dump(data: Any, explicit_start=True) -> str:
     )
 
 
-def safe_load(string) -> Dict:
+def new_safe_loader_with_include(import_path: str):
```
Comment (decentral1se) [inline src/molecule/util.py:258]: When is "new"? Today yes but tomorrow not. I would just go with `safe_loader_with_include`

**Review by decentral1se — COMMENTED** [review]
Overall I think this is a good change! Thanks. I am a bit wary to see re-implementation of a feature that YAML supports but since we do generation we need to do it ourselves? Maybe I don't understand it well and you can explain better. Is there anyway to avoid this? Perhaps not. In this case, then I think the implementation looks good. Left some minor comments.

Some documentation would be good.

File: src/molecule/model/schema_v3.py (hunk 2 above)
Comment (lebauce) [reply src/molecule/model/schema_v3.py:370]: Fixed

File: src/molecule/util.py (hunk 3 above)
Comment (lebauce) [reply src/molecule/util.py:258]: True :-) Fixed

File: src/molecule/config.py (hunk 1 above)
Comment (lebauce) [reply src/molecule/config.py:421]: Changed to pathlib.Path

**Comment by lebauce** [comment]
@decentral1se Thank you for your review. I'm not sure to understand what you mean by "re-implementation of a feature that YAML supports". AFAIK, including other files is not supported by YAML.

I tried to address your comments and added some doc.

**Comment by decentral1se** [comment]
> AFAIK, including other files is not supported by YAML.

Ahhhh, OK, I thought it did that.

Good then, looking good, thanks for thrashing out the comments.

Let's see what the rest say.

File: docs/examples.rst [hunk 4]
```diff
@@ -355,6 +358,30 @@ In this second example the actions `create`, `destroy`,
...
+.. code-block:: yaml
+
+    name: ansible
+    playbooks:
+      create: ../resources/playbooks/create.yml
+      destroy: ../resources/playbooks/destroy.yml
+      converge: ../resources/playbooks/converge.yml
+      prepare: ../resources/playbooks/prepare.yml
+
+The file and therefore be used in a scenario using:
```
Comment (larmbuster) [inline docs/examples.rst:376]: Should the "and" be a "can" instead?

Comment (lebauce) [reply docs/examples.rst:376]: Indeed. Fixed, thanks

**Review by decentral1se — APPROVED** [review]

**Review by larmbuster — APPROVED** [review]

**Comment by ssbarnea** [comment]
With all due respect, I am inclined to refuse this feature as it would introduce a serious amount of complexity and prevent us from doing other changes to molecule scenarios. At least now they are contained YAML files. Adding include opens a pandora box.

I would rather wait a little bit more and see if there are other ways to improve the user experience without this complexity.

**Comment by decentral1se** [comment]
> prevent us from doing other changes to molecule scenarios

Which changes?

> Adding include opens a pandora box

Having a limited form of this feature which is experiimental (weve done this before, remember parallel builds?) and not publicly documented seems to me a fine way to test this out! Some limited form of inclusion can be worth it, I think.

**Review by ssbarnea — CHANGES_REQUESTED** [review]
This will prevent use of simple (JSON/YAML) schema based validation for the file format and make almost impossible to integrate it with editors.

Molecule already supports layering config by loading the shared config, which should be enough for the moment (limited to only one shared config per repo).

I would be more inclined to support an approach that would allow to mention alternative file to be inherited from than adding a custom YAML extension.

I do want to keep using a simple YAML parsing, one without extensions.

That is why I would find a feature like `extends: filename` a more appropriate approach, especially as we can consider `.config/molecule/config.yml` to be the implicit value when the key is absent.

**Comment by lebauce** [comment]
@ssbarnea I think that adding YAML files at the parsing phase is somehow less complex that loading YAML files using variable interpolation (like for playbooks for instance). Allowing to specify external files (https://molecule.readthedocs.io/en/latest/examples.html#sharing-across-scenarios) requires adding specific code for every section.

**Comment by lebauce** [comment]
@ssbarnea Should I close the PR ?

**Pull request closed without merging**